snitch/meson.build

```diff
@@ -0,0 +1,70 @@
+version = meson.project_version()
+vers = version.split('.')
+
+git_hash = run_command(find_program('git'),'log','-1','--format=%h'
+                      ,check:true).stdout().strip()
+
+conf_data = configuration_data({
+  'PROJECT_VERSION' : version,
+  'PROJECT_VERSION_MAJOR' : vers[0],
+  'PROJECT_VERSION_MINOR' : vers[1],
+  'PROJECT_VERSION_PATCH' : vers[2],
+
+  'SNITCH_FULL_VERSION' : version + '.' + git_hash,
+
+  'SNITCH_MAX_TEST_CASES'        : get_option('max_test_cases'),
+  'SNITCH_MAX_NESTED_SECTIONS'   : get_option('max_nested_sections'),
+  'SNITCH_MAX_EXPR_LENGTH'       : get_option('max_expr_length'),
+  'SNITCH_MAX_MESSAGE_LENGTH'    : get_option('max_message_length'),
+  'SNITCH_MAX_TEST_NAME_LENGTH'  : get_option('max_test_name_length'),
+  'SNITCH_MAX_TAG_LENGTH'        : get_option('max_tag_length'),
+  'SNITCH_MAX_CAPTURES'          : get_option('max_captures'),
+  'SNITCH_MAX_CAPTURE_LENGTH'    : get_option('max_capture_length'),
+  'SNITCH_MAX_UNIQUE_TAGS'       : get_option('max_unique_tags'),
+  'SNITCH_MAX_COMMAND_LINE_ARGS' : get_option('max_command_line_args'),
+
+  'SNITCH_DEFINE_MAIN'           : get_option('define_main').to_int(),
+  'SNITCH_WITH_EXCEPTIONS'       : get_option('with_exceptions').to_int(),
+  'SNITCH_WITH_TIMINGS'          : get_option('with_timings').to_int(),
+  'SNITCH_WITH_SHORTHAND_MACROS' : get_option('with_shorthand_macros').to_int(),
+  'SNITCH_DEFAULT_WITH_COLOR'    : get_option('default_with_color').to_int(),
+  'SNITCH_CREATE_HEADER_ONLY'    : get_option('create_header_only').to_int(),
+  'SNITCH_CREATE_LIBRARY'        : get_option('create_library').to_int(),
+})
+
+conf_file = configure_file(
+  input: meson.project_source_root() / 'include/snitch/snitch_config.hpp.config',
+  output: 'snitch_config.hpp',
+  format: 'cmake',
+  configuration: conf_data,
+  install_dir: 'include/snitch'
+)
+
+snitch_config_cpp_args = [
```

Comment:
Why do you need these? Shouldn't `src/snitch.cpp` be getting this information from `snitch_config.hpp`?

Reply:
The CMake build adds these as `target_compile_definitions` here:
https://github.com/cschreib/snitch/blob/main/CMakeLists.txt#L57-L70

@cschreib - why are these needed?

This was my attempt to replicate all these defines in the meson target,
i.e. minimize the diff between cmake and meson generated targets
as seen by `meson introspect -i --targets build`

Reply:
The config header was added afterwards, and I never removed these compile definitions... I believe it should just work if they are removed.

I am not 100% sure what is the best way forward though. In principle in a pure CMake world, you could ditch the config header, and just rely on the compilation definitions of the imported target. This seems like the "modern" way. But it is purely a CMake thing; without the ability to read the installed CMake target files, this doesn't work. The config header is more old school, but is also more portable.

So I think I'd be OK with not setting the compile definitions, and rely on the config header instead. But I have limited experience in deploying software libraries, so I don't know if that is locking us in a corner?

Reply:
> The real old-school way :) is to have a config header for when the build system is used, and allow manually passing the defines for people who don't want to use the build system and don't have a config header.

Hmmm! We're sort of doing that with the current header; we allow the options to be overridden (they are defined only if missing). But we force inclusion of the config header, which isn't the same as what you describe. Actually, the code in the config header used to be located at the top of the main snitch header, with some sensible hard-coded defaults that could be overridden. 

> Regardless, the use of a config file vs. command line defines is primarily a matter of simplifying the compiler invocations, avoiding ARG_MAX, etc.

When building the library, yes. But I think they serve a different primary purpose when *using* the library: to avoid any mismatch in preprocessor definitions between when the library is built versus when it is used.

That is the main reason I created the config header in the first place.